paraview: add viskores in newer versions

kmorel opened the pull request:
Starting with ParaView 6.0, the VTK-m accelerator library changed to Viskores. Update the options to use the new accelerator library.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kmorel committed:
paraview: use viskores in newer versions

Starting with ParaView 6.0, the VTK-m accelerator library changed
to Viskores. Update the options to use the new accelerator library.

The approach has changed slightly. Earlier there was a use_vtkm
option to specify whether VTK-m should be compiled, and it always
compiled the version provided with ParaView. However, the API
for Viskores is stable, so use the external viskores package.

Co-authored-by: Ryan Krattiger <80296582+kwryankrattiger@users.noreply.github.com>

kmorel commented:
@kwryankrattiger Can you please review this?

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -152,18 +152,31 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
+        default="default",
+        multi=False,
+        values=("default", "on", "off"),
+        description="Build Viskores with ParaView by setting PARAVIEW_USE_VISKORES=ON,OFF."
```
Here and `when="@:5"` for the `use_vtkm` variant. This should make it possible to drop the added version constraint everywhere these are used.

```suggestion
        when="@6:",
        description="Build Viskores with ParaView by setting PARAVIEW_USE_VISKORES=ON,OFF."
```

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -152,18 +152,31 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
         ' "on" or "off" will always override the build_edition.',
     )
+    variant(
+        "use_viskores",
```
I am torn about adding this variant vs just renaming it to something like `accelerators` or making it implicit with `+cuda`, `+rocm`, `+fides` rather than having all of those be conflicts.

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -661,10 +677,17 @@ def use_x11():
         cmake_args.append("-DPARAVIEW_BUILD_SHARED_LIBS:BOOL=%s" % variant_bool("+shared"))
 
```
See previous suggestion to just drop the variant.

```suggestion
use_viskores = spec.satisfies("+cuda") or spec.satisfies("+rocm") or spec.satisfies("+fides")
```

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -175,6 +188,9 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("use_vtkm=on", when="@:5.3")
+    # After 5.13, ParaView switched from VTK-m to Viskores
+    conflicts("use_vtkm=on", when="@6.0:")
+    conflicts("use_viskores=on", when="@:5.13")
```
For ParaView 6.0 and later, we can use external Viskores. I would rather do that here if we can too. We will come back through on this as well and do external VTK, but that requires some TLC to the VTK package first.

kmorel replied:
I like the idea of using an external package. That is more in line with Spack and can simplify several things.

kwryankrattiger reviewed (commented):
Sorry for taking so long on this, I have been back and forth on how best to handle this going forward.

It may be better to just prune this variant at enable the accelerator implicitly. And then external Viskores would be really nice to not have to rebuild it from scratch if we don't have to, especially for the GPU stuff.

kmorel commented:
@kwryankrattiger I've updated this package to use an external viskores package. It also removes the `use_viskores` variant and instead implicitly adds the dependency when needed. I think that generally addresses your comments.

I have not extensively tested this, but the spec at least works.

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -308,13 +307,25 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
+    depends_on("viskores +rocm", when="@6: +rocm")
+    for target in ROCmPackage.amdgpu_targets:
+        depends_on(
+            f"viskores +rocm amdgpu_target={target}", when=f"@6: +rocm amdgpu_target={target}"
```
This can go inside the same loop as kokkkos.

kmorel replied:
> This can go inside the same loop as kokkkos.

I would rather keep these as separate loops. That keeps a separation of the code that is pre-v6 using an internal VTK-m and post-6 using Viskores and an external package.

kwryankrattiger replied:
Now that I think about it a bit more I agree, see my other comment to put the 6: stuff in a `when` block

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -308,13 +307,25 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
         depends_on(
             "kokkos@:3.7 +rocm amdgpu_target={0}".format(target),
-            when="+rocm amdgpu_target={0}".format(target),
+            when="@:5.13 +rocm amdgpu_target={0}".format(target),
```
You should be able to just use `@:5` here. I think that is more clear to read "All versions of 5 and older`

```suggestion
            when="@:5 +rocm amdgpu_target={0}".format(target),
```

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -661,11 +672,21 @@ def use_x11():
 
         # VTK-m added to ParaView in 5.3.0 and up
-        if spec.satisfies("@5.3.0:") and spec.variants["use_vtkm"].value != "default":
+        if spec.satisfies("@5.3.0:5.13") and spec.variants["use_vtkm"].value != "default":
```
```suggestion
        if spec.satisfies("@5.3.0:5") and spec.variants["use_vtkm"].value != "default":
```

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -308,13 +310,27 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
+
+    # For @6.0:, depend on external Viskores module when needed.
+    for variant in viskores_dependency_variants:
+        depends_on("viskores +vtktypes +64bitids", when="@6: +fides")
```
```suggestion
        depends_on("viskores +vtktypes +64bitids", when=f"@6: {variant}")
```

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -308,13 +310,27 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("zlib-api")
     depends_on("libcatalyst@2:", when="+libcatalyst")
-    depends_on("hip@5.2:", when="+rocm")
+
```
I think this entire section could be in a 

```
with when("@6:"):
    ....
```

kmorel replied:
Good suggestion. I implemented that.

kwryankrattiger commented on repos/spack_repo/builtin/packages/paraview/package.py:
```diff
@@ -743,7 +770,12 @@ def use_x11():
 
         if spec.satisfies("@5.11:"):
-            cmake_args.append("-DPARAVIEW_USE_HIP:BOOL=%s" % variant_bool("+rocm"))
+            if spec.satisfies("@6.0:"):
```
I think this would be easier to read as

```python
if spec.satisfies("@5.11:5"):
    cmake_args.append("-DPARAVIEW_USE_HIP:BOOL=%s" % variant_bool("+rocm"))
else:
    cmake_args.append("-DPARAVIEW_USE_KOKKOS:BOOL=%s" % variant_bool("+rocm"))

if "+rocm" in spec:
    if spec.satisfies("@6:"):
            cmake_args.append("-DPARAVIEW_KOKKOS_BACKEND:STRING=HIP")
    ...
```

kmorel replied:
Done.

kwryankrattiger commented:
Ported these changes to https://github.com/spack/spack-packages/pull/192 with other ParaView 6 Updates

kmorel commented:
Closing this PR. Changes are superseded by #192

bernhardkaindl commented:
kmorel wrote:
> Closing this PR. Changes are superseded by https://github.com/spack/spack-packages/pull/192
